File: app/src/main/java/eu/darken/bluemusic/main/core/service/modules/FakeDeviceConnectModule.java
package eu.darken.bluemusic.main.core.service.modules;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

import eu.darken.bluemusic.bluetooth.core.EventGenerator;
import eu.darken.bluemusic.bluetooth.core.FakeSpeakerDevice;
import eu.darken.bluemusic.bluetooth.core.SourceDevice;
import eu.darken.bluemusic.main.core.database.DeviceManager;
import eu.darken.bluemusic.main.core.database.ManagedDevice;
import eu.darken.bluemusic.main.core.service.ActionModule;
import eu.darken.bluemusic.main.core.service.BlueMusicServiceComponent;
import timber.log.Timber;

@BlueMusicServiceComponent.Scope
public class FakeDeviceConnectModule extends ActionModule {
    private final DeviceManager deviceManager;
    private final EventGenerator eventGenerator;

    @Inject
    FakeDeviceConnectModule(EventGenerator eventGenerator, DeviceManager deviceManager) {
        super();
        this.eventGenerator = eventGenerator;
        this.deviceManager = deviceManager;
    }

    @Override
    public void handle(ManagedDevice eventDevice, SourceDevice.Event event) {
        if (event.getType() != SourceDevice.Event.Type.DISCONNECTED) return;

        final Map<String, ManagedDevice> managed = deviceManager.observe().blockingFirst();
        ManagedDevice fakeSpeaker = managed.get(FakeSpeakerDevice.ADDR);
        Timber.d("FakeSpeaker: %s", fakeSpeaker);
        if (fakeSpeaker == null) return;

        Map<String, ManagedDevice> active = new HashMap<>();
        for (Map.Entry<String, ManagedDevice> entry : managed.entrySet()) {
            if (entry.getValue().isActive()) active.put(entry.getKey(), entry.getValue());
        }
        Timber.d("Active devices: %s", active);
        if (active.size() > 1 || active.size() == 1 && !active.containsKey(fakeSpeaker.getAddress())) return;

        Timber.d("Sending fake device connect event.");
        eventGenerator.send(fakeSpeaker.getSourceDevice(), SourceDevice.Event.Type.CONNECTED);
    }
}

File: app/src/main/java/eu/darken/bluemusic/main/core/service/modules/events/FakeDeviceConnectModule.kt
package eu.darken.bluemusic.main.core.service.modules.events

import dagger.Binds
import dagger.Module
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import dagger.multibindings.IntoSet
import eu.darken.bluemusic.bluetooth.core.SourceDevice
import eu.darken.bluemusic.bluetooth.core.speaker.FakeSpeakerDevice
import eu.darken.bluemusic.bluetooth.core.speaker.SpeakerDeviceProvider
import eu.darken.bluemusic.common.EventGenerator
import eu.darken.bluemusic.common.debug.logging.log
import eu.darken.bluemusic.common.debug.logging.logTag
import eu.darken.bluemusic.devices.core.DeviceRepo
import eu.darken.bluemusic.devices.core.ManagedDevice
import eu.darken.bluemusic.devices.core.currentDevices
import eu.darken.bluemusic.main.core.service.modules.EventModule
import kotlinx.coroutines.runBlocking
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class FakeDeviceConnectModule @Inject constructor(
    private val eventGenerator: EventGenerator,
    private val deviceRepo: DeviceRepo,
    private val speakerDeviceProvider: SpeakerDeviceProvider,
) : EventModule {

    override val priority: Int
        get() = 0

    override suspend fun handle(device: ManagedDevice, event: SourceDevice.Event) {
        if (event.type != SourceDevice.Event.Type.DISCONNECTED) return

        val managed = runBlocking { deviceRepo.currentDevices() }
        val fakeSpeaker = managed.singleOrNull { it.address == FakeSpeakerDevice.ADDRESS }
        log(TAG) { "FakeSpeaker: $fakeSpeaker" }
        if (fakeSpeaker == null) return

        val active = managed.filter { it.isActive }
        log(TAG) { "Active devices: $active" }
        if (active.size > 1 || active.size == 1 && !active.any { it.address == fakeSpeaker.address }) return

        log(TAG) { "Sending fake device connect event." }
        // Create a fake source device for the event
        val fakeSourceDevice = speakerDeviceProvider.getSpeaker()
        eventGenerator.send(fakeSourceDevice, SourceDevice.Event.Type.CONNECTED)
    }

    @Module @InstallIn(SingletonComponent::class)
    abstract class Mod {
        @Binds @IntoSet abstract fun bind(entry: FakeDeviceConnectModule): EventModule
    }

    companion object {
        private val TAG = logTag("FakeDeviceConnectModule")
    }
}